### PR TITLE
new test report generating trace

### PR DESCRIPTION
**Pull request by geez0219**
1. finished the feature to generate json summary for new test report

### CODE REVIEW

File: fastestimator/trace/io/test_report.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import os
+from typing import Callable, List, Union
+
+from fastestimator.trace.trace import Trace
+from fastestimator.util.data import Data
+from fastestimator.util.traceability_util import traceable
+from fastestimator.util.util import to_list
+from time import time
+from datetime import datetime
+import pdb
```
Comment (TortoiseHam) [inline fastestimator/trace/io/test_report.py:26]: pdb

File: fastestimator/trace/io/test_report.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        self.json_summary["execution_time(s)"] = time() - self.json_summary["execution_time(s)"]
+
+    def on_end(self, data: Data) -> None:
+        if self.json_output.endswith(".json"):
+            json_path = self.json_output
+        else:
+            json_path = os.path.join(self.json_output, "test_report.json")
+        with open(json_path, 'w') as fp:
+            json.dump(self.json_summary, fp, indent=4)
+        print("Saved test JSON report to {}".format(json_path))
```
Comment (TortoiseHam) [inline fastestimator/trace/io/test_report.py:89]: our code style calls for newline at end of files

File: fastestimator/trace/io/test_report.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+class TestReport(Trace):
+    """Automate testing and report generation.
+
+    Args:
+        test_descriptions: List of text-based descriptions.
+        test_criterias: List of test functions. Function input argument names needs to match keys from the data
+            dictionary.
+        test_title: Title of the test.
+        json_output: Path into which to write the output results JSON.
+        doc_output: Path into which to write the output QMS summary report (docx).
```
Comment (TortoiseHam) [inline fastestimator/trace/io/test_report.py:38]: unused doc parameter. and these two params should probably be consolidated into 1

File: fastestimator/trace/io/test_report.py (hunk 1 above)
Comment (geez0219) [reply fastestimator/trace/io/test_report.py]: fixed

File: fastestimator/trace/io/test_report.py (hunk 2 above)
Comment (geez0219) [reply fastestimator/trace/io/test_report.py]: fixed

File: fastestimator/trace/io/test_report.py (hunk 3 above)
Comment (geez0219) [reply fastestimator/trace/io/test_report.py]: fixed

**Pull request closed without merging**